### PR TITLE
improvement(header/p2p): get bestHead if timeout is reached

### PR DESCRIPTION
**Pull request by vgonkivs**
Closes https://github.com/celestiaorg/celestia-node/issues/1309

### CODE REVIEW

**Review by distractedm1nd — DISMISSED** [review]
nice!

File: header/p2p/exchange.go [hunk 1]
```diff
@@ -85,14 +85,15 @@ func (ex *Exchange) Head(ctx context.Context) (*header.ExtendedHeader, error) {
...
+LOOP:
 	for range ex.trustedPeers {
 		select {
 		case h := <-headerCh:
 			if h != nil {
 				result = append(result, h)
 			}
 		case <-ctx.Done():
-			return nil, ctx.Err()
+			break LOOP
```
Comment (Wondertan) [inline header/p2p/exchange.go:96]: We might also want to limit the Head call itself with a deadline. We don't do it where call `Head` in Syncer. The question is: should the user of the API do this himself, or should we do the limit for each Head call internally like we do it in the Server?

Comment (Wondertan) [reply header/p2p/exchange.go:96]: Other than this Q, LGTM, clean solution

Comment (vgonkivs) [reply header/p2p/exchange.go:96]: I would suggest to let users decide themselves. We have done it for Server side as this is a dos vector - to request 1000 range request that will block the whole service. But I don't think it's relevant for the client side.

Comment (renaynay) [reply header/p2p/exchange.go:96]: Let's put the above into an issue bc it applies to every client-side call @Wondertan 

At the moment, let's leave it for user/caller to decide until we see it causes issues / can be problematic.

Comment (Wondertan) [reply header/p2p/exchange.go:96]: @vgonkivs, pls make an issue

Comment (vgonkivs) [reply header/p2p/exchange.go:96]: https://github.com/celestiaorg/celestia-node/issues/1349

**Review by Wondertan — CHANGES_REQUESTED** [review]
Another thing I just realized about this PR. Instead of returning the expected context error on its cancellation, we may actually return a successful result. Such unexpected behaviors have to be explicitly documented, so pls, let's do this.

**Comment by Wondertan** [comment]
@vgonkivs, the comment should mainly be on the `Head`, because the behaviors of the method are different from what users usually expect. Note that this package will soon be used by other teams and such details will be useful for them

**Review by renaynay — APPROVED** [review]

**Review by Wondertan — APPROVED** [review]

**Pull request merged**